### DISCUSSION TITLE
Implement impersonating users & roles in App & API

### DISCUSSION THREAD

**Discussion by SeanDylanGoff** [directus/directus#19085] [Feature Requests]
### Summary

Admin users will be able to send the headers
`x-impersonate-user` and `x-impersonate-role`.
The API would then treat the user to have the given role / user id for all intents and purposes (permissions, presets etc).

There will also be a selector in the App for configuring if those headers are sent to the API. This selector should be easy to reach & refresh all data when changed

### Motivation

When developing a backend with multiple roles and role-specific presets mistakes can be easily overlooked, especially if users also have their own presets. I'm often confronted with problems that users have which are either role- or user specific. Having a way to impersonate users during QA and when responding to support tickets would make life a lot easier.
The current solution of either creating dummy users, or temporarily resetting user's passwords (making them reset it again after) is not good enough.

The extended idea of being able to impersonate roles in addition to users would for example allow verifying how a role change interacts with presets that the user already has without having to do the change.

### Detailed Design

The API has to handle the 2 new headers as follows:
- only `x-impersonate-user`: API treats requests as if logged in as the passed user, with the permissions of their role
- only `x-impersonate-role`: API treats requests as if the currently logged in (admin) user had the passed role
- both: API treats requests as if logged in with the passed user, and pretend that said user had the passed role

The App gets a new button visible on hover of the user icon (next to logout, only visible to admins). This icon triggers a modal for changing user & role, either using separate dropdowns or (probably better) using a more thought out UI element clearly distinguishing between the three use-cases.

### Must Have:
- Impersonate User
- Frictionless way to switch in App

### Should Have:
- Impersonate user with other role than they actually have
- Nice UX/UI for switching

### Could Have:
- Special treatment of activity tracking (making the activity entry "Admin User on behalf of John Doe" or "Admin User" instead of just "John Doe" when "Admin User" impersonates "John Doe")

### Won't Have:
-

### Drawbacks

- Implementation cost for the core requirements should be relatively small as far as I can determine from my limited knowledge of the codebase.
- Implementing it in user-space is AFAIK not possible without lots of ugly hacks
- Teaching impact should be zero as it's a feature that can be safely ignored by anyone who doesn't need it
- No idea on integrating with planned features
- Is not a breaking change, doesn't need any DB migrations

### Alternatives

Creating a dummy account per role (could be automated) and logging in with that. Kinda works, but is slow, hacky & doesn't solve the problem completely

### Adoption Strategy

Shouldn't affect anyone or anything in the ecosystem. It will just be an additional button that most users won't notice if they don't need it, and two headers that need to be documented.

**Comment by codihaus** [comment]
Would love to contribute the UXUI design for this case.
I will have my team to do a quick design this week, hope that we can have this come true!

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
@SeanDylanGoff Quick question! Seeing there's two, possibly 3 with #4959, variations of IDs to be passed in a header, should we go:

```
x-impersonate-user
x-impersonate-role
x-impersonate-token
```

or rather

```
x-impersonate-scope: role
x-impersonate-id: xxx
```

🤔

**Reply by rijkvanzanten** [reply]
Or potentially some sort of combined syntax like `x-impersonate: role 123`

**Reply by SeanDylanGoff** [reply]
What would token be?
I was thinking user and role instead of the second option, because it would allow one to impersonate a user, but pretend that user has a different role than they actually do.
But I'm not sure if that second use-case is even worth having.

**Reply by rijkvanzanten** [reply]
> What would token be?

As part of #4959, I've been thinking through a system where permissions are tied to policies rather than roles. Then each user would have a role + whatever additional policies you want to grant that one user. Each role would also be able to have one or more policies for standardized group policies.

In that setup, I was also thinking through the possibility of adding support for an API Token that's not explicitly linked to a _user_.

> I was thinking user and role instead of the second option, because it would allow one to impersonate a user, but pretend that user has a different role than they actually do.
> But I'm not sure if that second use-case is even worth having.

Ah right! I guess a single header can have whatever string value we want, so we could also go with a `x-impersonate: role 123, user 567` 🤔 HTTP headers are all over the place, so I'm not entirely sure what's the common or standard way of doing this

**Reply by SeanDylanGoff** [reply]
ah, i see. i'll read #4959 as soon as i have time. already loving it just by reading the first few sentences though. i think either way is fine (i just first misunderstood and thought you forgot about the possibility of specifying both a user and a role). 
however, personally i think separate headers are more explicit and i dislike string manipulation when passing data. so if it was up to me i'd either go for separate headers, or one header with a JSON payload like the "nel" header

**Comment by ricardovigatti** [comment]
Might be useful an environment variable or in-app configuration to control whether this function is enabled in server.

**Comment by rijkvanzanten** [comment]
Love this!

Quick couple of questions that come to mind:

- When modifying data when impersonating another user, who is the one that's tracked in accountability and revisions?
- Should we add a dedicated "impersonated_by" flag in accountability tracking to counter that?
- When impersonating a user, what is the value for `$CURRENT_USER` in validation/permission rules?
- Should we add an environment variable to enable/disable impersonation altogether?
- Note rather than question: this should be an admin-only feature

**Reply by ricardovigatti** [reply]
**When modifying data when impersonating another user, who is the one that's tracked in accountability and revisions?**
Both.

**Should we add a dedicated "impersonated_by" flag in accountability tracking to counter that?**
Yes.

**When impersonating a user, what is the value for $CURRENT_USER in validation/permission rules?**
The user who has been impersonated.

**Should we add an environment variable to enable/disable impersonation altogether?**
Yes.

**Note rather than question: this should be an admin-only feature**
Ok.

**Comment by bogdan-nita** [comment]
any updates on this?

**Comment by SeanDylanGoff** [comment]
We're currently working on a project with directus 11 which would benefit massively from this. However, doing so I realized that there is a small addition that would make this even more useful: Allow impersonating for non-admin users using custim verification logic. This could be implemented as a new API hook, on which a flow or extension can either allow, disallow or overwrite impersonation based on:

- accountability
- requested impersonation (the two headers)
- request payload/headers

I see this discussion got a lot of upvotes - but AFAIK there are no concrete plans to implement this yet, right? We might opt to either implement this ourselves, but since at least 66 other devs could benefit from this, it'd be great to agree on how this should be done, so it can either be merged or at least open to everyone as a fork.

**Reply by rijkvanzanten** [reply]
> but AFAIK there are no concrete plans to implement this yet, right? 

That's correct. I gotta pick my battles wisely, and while this is high on my personal would-like-to-have list, I have some other things at the top of the list first (mostly around scalability and stability)

**Comment by zsaraf** [comment]
Hello, any updates on this? This feature feels quite necessary for production environments when dealing with important clients!

**Reply by rijkvanzanten** [reply]
> This feature feels quite necessary for production environments when dealing with important clients!

Mind elaborating on why this is the case for your use cases? Wanna make sure that whatever the final solution looks like is relevant for your use case as well

**Reply by zsaraf** [reply]
Hi @rijkvanzanten 

Thanks for your response. Here are a few points on why this feature is important for us:

**Client Experience Verification**
When dealing with important clients, we often need to ensure that they’re seeing exactly what they should be seeing. Relying solely on permissions and configurations to display the correct data can be risky—it's hard to just "trust" that everything is set up perfectly. Being able to impersonate a user (or user + role) allows us to verify their exact experience quickly, giving us peace of mind before we finalize anything or after onboarding them.

**Bug Reproduction**
If a client reports an issue and we can't reproduce it, we're left with suboptimal workarounds like:

- Creating a separate user and trying to mirror their permissions and configurations, which is error-prone and time-consuming.
- Resetting their password temporarily, which disrupts their access and can feel unprofessional.

Being able to impersonate them directly simplifies reproducing bugs and ensures we're diagnosing the issue in the exact environment they're experiencing.

**Admin Access Context**
Since admins have access to all data, impersonation shouldn't introduce new risks in terms of data visibility. It simply enhances our ability to see things through the user's perspective without friction.

Overall, this feature would streamline our workflow, reduce guesswork, and improve the quality of our support and client interactions.

**Reply by rijkvanzanten** [reply]
Thanks! These are very helpful insights 🚀

**Reply by AndreKR** [reply]
We're also missing this a lot. Currently our admin users all have a separate second account (for which we have to pay I believe) that they can assign various roles to, so they can verify that the roles are seeing/able to update exactly what they need to. This is also a blocker for fully switching to SSO because with SSO we couldn't have second accounts.

**Comment by Svashtar** [comment]
Plus 1, +1
